src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -360,12 +375,39 @@ private bool TryHandleAsKeySegment(string segmentText)
             KeySegment previousKeySegment = this.FindPreviousKeySegment();
 
             KeySegment keySegment;
+
+            bool trybindingEscapeFunction = false;
+
+            if (segmentText.Length > 0 && segmentText[segmentText.Length - 1] == ':' && previous.EdmType != null)
+            {
+                IEdmFunction function = configuration.Model.FindBoundOperations(previous.EdmType).OfType<IEdmFunction>().FirstOrDefault(f => IsUrlEscapeFunction(configuration.Model, f));
+                if (function != null)
+                {
+                    trybindingEscapeFunction = true;
+                    segmentText = segmentText.Substring(0, segmentText.Length - 1);
+
+                    // If nothing to bind for key then bind the escape function directly. 
+                    if (segmentText.Length == 0)
+                    {
+                        return this.TryBindEscapeFunction();

```

Comment:
>return [](start = 24, length = 6)

is this the right semantics, to base the return of TryHandleAsKeySegment on the return from TryBindEscapeFunction? don't you know at this point that TryBindEscapeFunction will return true (since you already verified the escape function), and wouldn't it be an error (not false for evaluating this segment as a key) if TryBindEscapeFunction returned false?

Reply:
It can throw an error even if there is an escape function bound to this type. It will never return false as of now. More details in comments, let me know how you would like me to proceed. 
I changed the code as I thought was the best possible way to handle it. 
PS: I dislike the int passed to tell how many segments to remove from the list.

---
In reply to: [366150964](https://github.com/OData/odata.net/pull/1621#discussion_r366150964) [](ancestors = 366150964)